// jest-mock-extended/lib/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.calledWithFn = exports.stub = exports.mockFn = exports.mockReset = exports.mockClear = exports.mockDeep = exports.JestMockExtended = exports.mock = void 0;
var Mock_1 = require("./Mock");
Object.defineProperty(exports, "mock", { enumerable: true, get: function () { return __importDefault(Mock_1).default; } });
Object.defineProperty(exports, "JestMockExtended", { enumerable: true, get: function () { return Mock_1.JestMockExtended; } });
Object.defineProperty(exports, "mockDeep", { enumerable: true, get: function () { return Mock_1.mockDeep; } });
Object.defineProperty(exports, "mockClear", { enumerable: true, get: function () { return Mock_1.mockClear; } });
Object.defineProperty(exports, "mockReset", { enumerable: true, get: function () { return Mock_1.mockReset; } });
Object.defineProperty(exports, "mockFn", { enumerable: true, get: function () { return Mock_1.mockFn; } });
Object.defineProperty(exports, "stub", { enumerable: true, get: function () { return Mock_1.stub; } });
var CalledWithFn_1 = require("./CalledWithFn");
Object.defineProperty(exports, "calledWithFn", { enumerable: true, get: function () { return __importDefault(CalledWithFn_1).default; } });
__exportStar(require("./Matchers"), exports);
